### ISSUE TITLE
Labelled tuple respect multiline layout like type

### ISSUE THREAD

**Issue by josh-degraw** [prettier/prettier#10049]
I tried seeing if something already exists to address this, but couldn't find any, please feel free to let me know if such an issue already exists.

Basically, I would expect that a labelled tuple type would behave similarly to an object literal type in formatting.

For example, with this config
<details>
<summary> `.prettierrc` </summary>
<p>

```json
{
  "printWidth": 80,
  "tabWidth": 2,
  "semi": false,
  "singleQuote": true,
  "quoteProps": "as-needed",
  "jsxSingleQuote": false,
  "trailingComma": "none",
  "bracketSpacing": true,
  "jsxBracketSameLine": false,
  "arrowParens": "avoid",
  "overrides": [
    {
      "files": [
        "*.ts",
        "*.tsx"
      ],
      "options": {
        "parser": "typescript",
        "printWidth": 120
      }
    }
  ]
}
```
</p>
</details> 

If I have an object `type` definition like this:
```ts
// Both of these layouts are preserved when formatting
type UseConnectedLoaderResult = {
  loaderElement: React.ReactNode
  showLoader: boolean
  queryStatus: QueryStatus
}
type UseConnectedLoaderResult = { loaderElement: React.ReactNode; showLoader: boolean; queryStatus: QueryStatus }
```
Prettier will respect the layout either way, if I put it all on one line or put each property on one line.

But if I have the exact same layout with a labelled tuple, since it doesn't reach the `printWidth` in my config, it forces it on one line like so 
```ts
type UseConnectedLoaderResult = [loaderElement: React.ReactNode, showLoader: boolean, queryStatus: QueryStatus]
```
If I change the name to be longer than the max `printWidth`, it does format it how I would expect:
```ts
type UseConnectedLoaderResultWithALongerName = [
  loaderElement: React.ReactNode,
  showLoader: boolean,
  queryStatus: QueryStatus
]
```

Basically what I would want is to either have the labelled tuple formatting work the same way as type literals, or allow a way to specify that that's my desired behavior.

**Comment by sosukesuzuki** [comment]
I think this makes sense. Just apply Prettier's [multiline object formatting policy](https://prettier.io/docs/en/rationale.html#multi-line-objects) to TS labeled types.

**Comment by thorn0** [comment]
Looks like it makes sense to always split labelled tuples to multiple lines, doesn't it?

Anyway, preserving line breaks isn't a solution. See the [option philosophy](https://prettier.io/docs/en/option-philosophy.html) page:

>> Please note that as option requests are out of scope for Prettier, they will be closed without discussion. The same applies to requests to preserve elements of input formatting (e.g. line breaks) since that’s nothing else but an option in disguise with all the downsides of “real” options.

That said, I'd like to expand on some points:

> If I have an object type definition ... Prettier will respect the layout either way, if I put it all on one line or put each property on one line

This is a workaround, not a feature. It was implemented for object literals because at the time a good heuristic wasn't found and an urgent fix was needed. When the support for TS was implemented, the same code was used for object types. But as a general strategy, Prettier avoids irreversible behavior like that, so the team is still looking for heuristics that would allow either to remove this behavior completely or at least to reduce the number of situations where it's applied.

What does "irreversible" mean? Once an object literal becomes multiline, Prettier won't collapse it back. If in Prettier-formatted code, we add a property to an object literal, run Prettier, then change our mind, remove the added property, and then run Prettier again, we might end up with a formatting not identical to the initial one. This useless change might even get included in a commit, which is exactly the kind of situation Prettier was created to prevent.

Also, note that setting `printWidth` to high values is not recommended exactly because of this kind of undesirable formatting. See https://prettier.io/docs/en/options.html#print-width

**Comment by sosukesuzuki** [comment]
Okay, you are right.

> Looks like it makes sense to always split labelled tuples to multiple lines, doesn't it?

However I think it is overkill. I'll look for heuristics.
